### PR TITLE
Issue #26: Add a test for rule loading from a dot-directory

### PR DESCRIPTION
**Pull request by mdimjasevic**
This confirms that "Would it look at `.foo/bar.yaml` if `RUNTIME_IGNOREDOTFILES` is enabled?" has a positive answer. Ratelimit does the same.

### CODE REVIEW

File: test/Fencer/Rules/Test.hs [hunk 1]
```diff
@@ -26,20 +26,38 @@ tests = testGroup "Rule tests"
   [ test_rulesLoadRulesYaml
   , test_rulesLoadRulesNonYaml
   , test_rulesLoadRulesRecursively
+  , test_rulesLoadRulesDotDirectory
   ]
 
+-- | A helper function for loading rules and making sure they are as
+-- expected.
+loadRules :: String -> Bool -> IO ()
```
Comment (neongreen) [inline test/Fencer/Rules/Test.hs]: `String` and `Bool` should be named arguments.

File: test/Fencer/Rules/Test.hs [hunk 2]
```diff
@@ -26,20 +26,38 @@ tests = testGroup "Rule tests"
...
-        (sortOn domainDefinitionId [domain1, domain2])
-        (sortOn domainDefinitionId definitions)
+    loadRules "fencer-config" True
+
+-- | test that 'loadRulesFromDirectory' loads rules from a
+-- dot-directory when dot-files should be ignored.
+test_rulesLoadRulesDotDirectory :: TestTree
+test_rulesLoadRulesDotDirectory =
+  testCase "Rules are loaded from a dot-directory" $
+    loadRules ".fencer-config" True
```
Comment (neongreen) [inline test/Fencer/Rules/Test.hs]: Given our implementation of `loadRulesFromDirectory`, it's possible that it won't apply `#ignoreDotFiles` to the passed directory – only to its subdirectories. So the test should use a different directory structure:

```
<tempdir>
    .subdir
        config.yaml
```

File: test/Fencer/Rules/Test.hs (hunk 1 above)
Comment (neongreen) [reply test/Fencer/Rules/Test.hs]: Also:

```suggestion
loadRules :: String -> Bool -> Assertion
```

Comment (neongreen) [reply test/Fencer/Rules/Test.hs]: Also, what about generalizing it a bit?

```haskell
-- | Create given directory structure and check that 'loadRulesFromDirectory'
-- produces expected result.
expectLoadRules
  :: "ignoreDotFiles" :! Bool
  -> "files" :! [(FilePath, Text)]
  -> "result" :! [DomainDefinition]
  -> Assertion
```

Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: This looks good! I'll adopt it.

Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: Done.

File: test/Fencer/Rules/Test.hs (hunk 2 above)
Comment (mdimjasevic) [reply test/Fencer/Rules/Test.hs]: I updated this per your comment.

**Comment by neongreen** [comment]
How did you test `lyft/ratelimit`s behavior? I tested it too and it seems to ignore dot-directories:

```
$ tree -a examples

examples
├── root -> root2
└── root2
    └── ratelimit
        └── config
            └── .blah
                └── config.yaml
```

```
$ RUNTIME_IGNOREDOTFILES=false \
  RUNTIME_ROOT=examples/root \
  RUNTIME_SUBDIRECTORY=ratelimit \
  LOG_LEVEL=DEBUG \
  bin/ratelimit

DEBU[0000] runtime changed. loading new snapshot at examples/root/ratelimit
DEBU[0000] runtime: processing examples/root/ratelimit
DEBU[0000] runtime: processing examples/root/ratelimit/config
DEBU[0000] runtime: processing examples/root/ratelimit/config/.blah
DEBU[0000] runtime: processing examples/root/ratelimit/config/.blah/config.yaml
DEBU[0000] runtime: adding key=config..blah.config.yaml value=---
domain: mongo_cps2
descriptors:
  - key: database
    value: default
    rate_limit:
      unit: hour
      requests_per_unit: 4
 uint=false
WARN[0000] connecting to redis on unix /var/run/nutcracker/ratelimit.sock with pool size 10
```

```
$ RUNTIME_IGNOREDOTFILES=true \
  RUNTIME_ROOT=examples/root \
  RUNTIME_SUBDIRECTORY=ratelimit \
  LOG_LEVEL=DEBUG \
  bin/ratelimit

DEBU[0000] runtime changed. loading new snapshot at examples/root/ratelimit
DEBU[0000] runtime: processing examples/root/ratelimit
DEBU[0000] runtime: processing examples/root/ratelimit/config
DEBU[0000] runtime: processing examples/root/ratelimit/config/.blah
WARN[0000] connecting to redis on unix /var/run/nutcracker/ratelimit.sock with pool size 10
```

**Comment by mdimjasevic** [comment]
> How did you test `lyft/ratelimit`s behavior? I tested it too and it seems to ignore dot-directories:

I guess I had an example where the top-level directory given by the `RUNTIME_ROOT` environment variable had a name starting with a dot, which `lyft/ratelimit` happily loads rules from regardless of the `RUNTIME_IGNOREDOTFILES` flag.

Ok, I'll update this PR per your observation.

**Comment by mdimjasevic** [comment]
@neongreen , at least one other PR of mine depends on this one so it'd be great if you can take a look again.

File: lib/Fencer/Rules.hs [hunk 3]
```diff
@@ -26,22 +26,28 @@ import Fencer.Types
 --
 -- Throws an exception for unparseable or unreadable files.
```
Comment (neongreen) [inline lib/Fencer/Rules.hs:30]: Ok, so this skips all dot-directories including `subDirectory` and below. Right? If yes, can you add it to the haddock?

**Review by neongreen — APPROVED** [review]

File: lib/Fencer/Rules.hs (hunk 3 above)
Comment (mdimjasevic) [reply lib/Fencer/Rules.hs:30]: That's right. I expanded the documentation for this.

**Comment by neongreen** [comment]
In case you're waiting on me – this can be merged.

**Pull request merged**